pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -928,6 +943,27 @@ public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
         }
     }
 
+    private int removeExpiredMessagesFromQueue(Set<MessageIdImpl> messageIds) {
+        int messagesFromQueue = 0;
+        for (Message message : incomingMessages) {
+            MessageIdImpl messageId = (MessageIdImpl) message.getMessageId();
+            if (messageId instanceof BatchMessageIdImpl) {
+                // messageIds contain MessageIdImpl, not BatchMessageIdImpl
+                messageId = new MessageIdImpl(messageId.getLedgerId(), messageId.getEntryId(),
+                        getPartitionIndex());
+            }
+            if (messageIds.contains(messageId)) {
```

Comment:
The queue is being accessed concurrently, the item we peek may not be the one we end up popping.
If we were to use removeIf we can't really know if we actually managed to remove the item or someone else did. Even so, removeIf requires an iterator I believe.